bin/cylc-check-software

```diff
@@ -16,84 +16,238 @@
+            ret.append(sub_version)
+    return tuple(ret)
+
+
+def string_ver(version_tuple):
+    return '.'.join(str(x) for x in version_tuple)
+
+
+def shell_align_write(one_delimiter, left_msg, right_msg):
+    """Write two messages aligned with the terminal edges, separated by a
+       given delimiter, with a minimum separation of two characters."""
+    gap = output_width() - len(left_msg) - len(right_msg)
+    if gap >= 2:
+        sys.stdout.write(left_msg + one_delimiter * gap + right_msg + '\n')
+        return True
+    else:
+        return False
+
+
+def shell_centre_write(prepend_newline, *args):
+    """Write one or more lines of text centrally in the terminal."""
+    if prepend_newline:
+        sys.stdout.write('\n')
+    for msg_line in args:
+        spacing = int(round(output_width() - len(msg_line)) / 2) * ' '
+        sys.stdout.write(spacing + msg_line + spacing + '\n')
+    return
+
+
+def check_py_ver(min_ver, max_ver=None):
+    """Check if a version of Python within a specified range is installed."""
+    if max_ver:
+        msg = 'python (%s+, <%s)' % (string_ver(min_ver), string_ver(max_ver))
+    else:
+        msg = 'python (%s+)' % string_ver(min_ver)
+    version = sys.version_info
+    ret = (version >= min_ver and (not max_ver or version < max_ver))
+    shell_align_write('.', msg, '%s (%s)' % (MINVER_MET_MSG if ret else
+                      MINVER_NOTMET_MSG, string_ver(version)))
+    return ret
+
+
+def check_py_module_ver(module, min_ver):
+    """Check if a minimum version of a Python module is installed."""
+    msg = 'Python:%s (%s)' % (module, string_ver(min_ver) + '+' if
+                              min_ver is not None else 'none')
+    try:
+        if module == 'pygtk':
+            import gtk
+            module_version = '.'.join(map(str, gtk.pygtk_version))
+        else:
+            imported_module = __import__(module)
+            module_version = imported_module.__version__
+    except ImportError:
+        res = [NOTFOUND_MSG, False]
+    try:
+        if min_ver is None:
+            res = ['%s (%s)' % (FOUND_NOVER_MSG, module_version), True]
+        elif parse_version(module_version) >= min_ver:
+            res = ['%s (%s)' % (MINVER_MET_MSG, module_version), True]
+        else:
+            res = ['%s (%s)' % (MINVER_NOTMET_MSG, module_version), False]
+    except AttributeError:
+        res = [FOUND_UNKNOWNVER_MSG, False]
+    shell_align_write('.', msg, res[0])
+    return res[1]
+
+
+def kpsepath_search(tex_module):
+    """Print outcome of local TeX module search using 'kpsepath' command."""
+    # Test line-count; 'locate' output can be >> for acc. string-length test
+    command = ("kpsepath tex | locate '*tex*" + tex_module + "*' | wc -l")
+    msg = 'TeX:%s (none)' % tex_module
+    try:
+        init = Popen(command, shell=True, stdout=PIPE, stderr=PIPE)
+        init.wait()
+        final_command = ['test', init.communicate()[0].strip(), '-ne', '0']
```

Comment:
No need to (in fact it may be advisable not to) use subprocess `Popen.wait()` before `Popen.communicate()` - the latter will "wait" anyway. 
 * https://docs.python.org/2/library/subprocess.html#subprocess.Popen.wait 
 * https://docs.python.org/2/library/subprocess.html#subprocess.Popen.communicate

Reply:
Indeed! Well spotted - I do not yet have much experience using the `Popen` class so I am not surprised I have not quite used it correctly. Will amend that now.